### PR TITLE
Implement Stage instances

### PR DESCRIPTION
**Pull request by DRSchlaubi**
Reference: https://github.com/discord/discord-api-docs/pull/2898/

### CODE REVIEW

File: core/src/main/kotlin/behavior/StageInstanceBehavior.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    object : StageInstanceBehavior {
+        override val channelId: Snowflake
+            get() = channelId
+        override val kord: Kord
+            get() = kord
+        override val id: Snowflake
+            get() = id
+        override val supplier: EntitySupplier
+            get() = supplier
+    }
```
Comment (BartArys) [inline core/src/main/kotlin/behavior/StageInstanceBehavior.kt:46]: Consider adding a meaningful toString

File: core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt [hunk 2]
```diff
@@ -69,6 +73,20 @@ suspend fun StageChannelBehavior.edit(builder: StageVoiceChannelModifyBuilder.()
...
+    return StageInstance(data, kord, supplier)
+}
+
+suspend fun StageChannelBehavior.getStageInstance(): StageInstance {
+    val instance = kord.rest.stageInstance.getStageInstance(id)
+    val data = StageInstanceData.from(instance)
+
+    return StageInstance(data, kord, supplier)
+}
+
```
Comment (BartArys) [inline core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt:89]: These shouldn't be extension functions. Generally online inline functions are added as extensions for interfaces.

File: core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt [hunk 3]
```diff
@@ -69,6 +73,20 @@ suspend fun StageChannelBehavior.edit(builder: StageVoiceChannelModifyBuilder.()
...
+suspend fun StageChannelBehavior.getStageInstance(): StageInstance {
+    val instance = kord.rest.stageInstance.getStageInstance(id)
+    val data = StageInstanceData.from(instance)
+
+    return StageInstance(data, kord, supplier)
+}
```
Comment (BartArys) [inline core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt:88]: Consider adding a safe nullable version.

**Review by BartArys — CHANGES_REQUESTED** [review]

File: core/src/main/kotlin/behavior/StageInstanceBehavior.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        return StageInstance(data, kord, supplier)
+    }
+
+    suspend fun asStageInstance(): StageInstance {
+        val channel = kord.rest.stageInstance.getStageInstance(channelId)
+        val data = StageInstanceData.from(channel)
+
+        return StageInstance(data, kord, supplier)
+    }
```
Comment (BartArys) [inline core/src/main/kotlin/behavior/StageInstanceBehavior.kt:30]: Delegate to supplier instead

File: core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt [hunk 5]
```diff
@@ -30,6 +32,20 @@ interface StageChannelBehavior : BaseVoiceChannelBehavior {
...
+
+        return StageInstance(data, kord, supplier)
+    }
+
+    suspend fun getStageInstanceOrNull(): StageInstance? =
+        withStrategy(EntitySupplyStrategy.rest).supplier.getStageInstanceOrNull(id)
+
+    suspend fun getStageInstance(): StageInstance =
+        withStrategy(EntitySupplyStrategy.rest).supplier.getStageInstance(id)
+
```
Comment (BartArys) [inline core/src/main/kotlin/behavior/channel/StageChannelBehavior.kt:48]: delegate this to the supplier of the behavior instead

File: core/src/samples/kotlin/PingBot.kt [hunk 6]
```diff
@@ -8,6 +12,17 @@ suspend fun main(args: Array<String>) {
...
+            val topic = message.content.substringAfter("topic").trim()
+            val channelId = member?.getVoiceState()!!.channelId!!
+            val channel = getGuild()!!.getChannelOf<StageChannel>(channelId)
+
+            if (topic.isNotEmpty()) {
+                channel.createStageInstance(topic)
+            } else {
+                message.channel.createMessage("Topic: ${channel.getStageInstance().topic}")
+            }
+        }
```
Comment (BartArys) [inline core/src/samples/kotlin/PingBot.kt:25]: Why was this added?

Comment (DRSchlaubi) [reply core/src/samples/kotlin/PingBot.kt:25]: Because I clicked the wrong button

**Review by BartArys — APPROVED** [review]
